pydantic/json.py

```diff
@@ -35,7 +35,8 @@ def decimal_encoder(dec_value: Decimal) -> Union[int, float]:
     >>> decimal_encoder(Decimal("1"))
     1
     """
-    if dec_value.as_tuple().exponent >= 0:
+    exponent = dec_value.as_tuple().exponent
+    if isinstance(exponent, int) and exponent >= 0:
         return int(dec_value)
```

Comment:
I can change `isinstance` to use try/exception for performance.

Reply:
I think this is fine, we'll need to change this code anyway.